src/bridge/DelayBuffer.sol

```diff
@@ -0,0 +1,347 @@
+            syncProof.delayedMessage.timestamp
+        );
+    }
+
+    /// @dev    Calculates the margin a sequenced message is below the delay threshold
+    ///         defining a `sync validity` window during which no delay proofs are required.
+    /// @notice Updates the time / block until no delay proofs are required.
+    /// @param blockNumber The block number when the synced message was created.
+    /// @param timestamp The timestamp when the synced message was created.
+    function updateSyncValidity(
+        BufferData storage self,
+        BufferConfig memory bufferConfig,
+        uint64 blockNumber,
+        uint64 timestamp
+    ) internal {
+        // saturating at uint64 max gracefully handles large threshold settings
+        self.syncExpiryBlockNumber = bufferConfig.thresholdBlocks > type(uint64).max - blockNumber
+            ? type(uint64).max
+            : blockNumber + bufferConfig.thresholdBlocks;
+        self.syncExpiryTimestamp = bufferConfig.thresholdSeconds > type(uint64).max - timestamp
+            ? type(uint64).max
+            : timestamp + bufferConfig.thresholdSeconds;
+    }
+
+    function isOnTime(
+        uint64 blockNumber,
+        uint64 timestamp,
+        uint64 thresholdBlocks,
+        uint64 thresholdSeconds
+    ) internal view returns (bool) {
+        return ((uint64(block.number) - blockNumber <= thresholdBlocks) &&
+            (uint64(block.timestamp) - timestamp <= thresholdSeconds));
+    }
+
+    /// @dev    Depletion is rate limited to allow the sequencer to recover properly from outages.
+    ///         In the event the batch poster is offline for X hours and Y blocks, when is comes back
+    ///         online, and sequences delayed messages, the buffer will not be immediately depleted by
+    ///         X hours and Y blocks. Instead, it will be depleted by the time / blocks elapsed in the
+    ///         delayed message queue. The buffer saturates at the threshold which allows recovery margin.
+    /// @notice Decrements the delay buffer saturating at the threshold
+    /// @param start The beginning reference point (prev delay)
+    /// @param end The ending reference point (current message)
+    /// @param delay The delay to be applied (prev delay)
+    /// @param threshold The threshold to saturate at
+    /// @param buffer The buffer to be decremented
+    function deplete(
+        uint64 start,
+        uint64 end,
+        uint64 delay,
+        uint64 threshold,
+        uint64 buffer
+    ) internal pure returns (uint64) {
+        uint64 elapsed = end > start ? end - start : 0;
+        uint64 unexpectedDelay = delay > threshold ? delay - threshold : 0;
+        uint64 decrease = unexpectedDelay > elapsed ? elapsed : unexpectedDelay;
+        // decrease the buffer saturating at zero
+        buffer = buffer > decrease ? buffer - decrease : 0;
+        // saturate at threshold
+        buffer = buffer > threshold ? buffer : threshold;
```

Comment:
i don't really understand saturating at the threshold

Reply:
what would happen if the buffer is allowed to deplete to zero?

L1->L2 messages could be force included immediately. Force inclusion increments the sequencerNumber, so the sequencer attempting to include a batch may have their batches revert due to a sequencerNumber mismatch. The sequencer would need to post batches setting the sequencerNumber to type(uint256).max to bypass the check.

Further, the sequencer could not give out any preconfirmations when it recovers (back online after outage) since there is no delay on force inclusions and therefore L2 reorgs are a concern. The sequencer would be required to sequence exclusively delayed messages until the buffer replenishes sufficiently (could be hours to days). Arguably this is a good feature.

----

Alternatively, as written, the buffer minimum is the threshold (perhaps 30 min, or a couple hours). Then when the sequencer recovers is has some margin to issue preconfirmations immediately. This also means that L1->L2 messages would be required to wait atleast the threshold at all times.

For the requirements of reasonable challenge times for L3 fraud proof games, this means that 50 steps of bisection would require 50*threshold more time compared to the L2 fraud proof game on L1. So for reasonable threshold ranges (30 min - 2 hours), than implies an additional challenge time for L3s of a few days.

---

What do you think? Saturation at the threshold allows the sequencer to recover "smoothly". The trade off is minimum force inclusion time which doesn't have large impact on the delays of L3 fraud proof games. On the otherhand, allowing the buffer to drop to zero would mean minimal delay for L3 fraud proofs, but somewhat more awkward recovery for the Sequencer.